### ISSUE TITLE
Training simple_object_detector in Python and then loading it in C++ problem

### ISSUE THREAD

**Issue by florian** [davisking/dlib#671]
I'm training a `simple_object_detector` using dlib's Python API:

```python
options = dlib.simple_object_detector_training_options()
options.add_left_right_image_flips = False
options.C = 5
options.num_threads = 3
options.be_verbose = True

dlib.train_simple_object_detector("data.xml", "detector.svm", options)
```

If I then load this detector in Python using `dlib.simple_object_detector("detector.svm")` and try to evaluate it on a dataset, it works really really well. The problem is that we need to use the detector in C++, and there it barely works.

The `detector.svm` is loaded like this in C++:

```c++
typedef scan_fhog_pyramid<pyramid_down<6> > image_scanner_type; 
image_scanner_type scanner;

object_detector<image_scanner_type> detector;
deserialize("face_detector.svm") >> detector;
```

It only works on very few of our test images (like 5 out of 100), whereas in Python we get most of them correctly (like 90-95 out of 100). I assume that the .svm file is not loaded correctly in C++.

If I load the .svm file in Python using `dlib.fhog_object_detector("detector.svm")`, then I actually get the same behaviour as in C++: It only works for the same 5 images. But this is a bit confusing, because the detector wasn't trained as an fhog object detector, but the method also doesn't throw an error.

There is another weird behaviour: The bounding box found in C++ seems to be different from the one found in Python. Here's the bounding box we get for one of the test images in Python:

![py-frame35](https://user-images.githubusercontent.com/1564556/27741996-51ea02b4-5db7-11e7-834b-df3fbb3918fb.png)

And here is the bounding box generated by the exact same .svm file in C++ (The screenshot is a bit bigger, but of course both programs run on the same image):

![cpp-frame35](https://user-images.githubusercontent.com/1564556/27741739-67820672-5db6-11e7-97dd-ccd320d5f49e.jpg)

It seems like in C++ the detector tries a much larger sliding window, and therefore doesn't find any objects if they are smaller.

This is the C++ code that's being used for giving the OpenCV image to the dlib detector:

```c++
cv::Mat imgForDlib = cv_ptr->image.clone();
dlib::cv_image<dblib::bgr_pixel> cimg(imgForDlib);
std::vector<dlib::rectangle> objects = detector(cimg);
```

Is that detector not loaded correctly in C++? Or is it a problem with the image datatype that we give to dlib?
We would greatly appreciate any help!

**Comment by florian** [comment]
In the Python training part this is printed after training is done:

```
Trained with sliding window 82 pixels wide by 78 pixels tall.
Upsampled images 2 times to allow detection of small boxes.
```

I'm not sure if this helps with the issue. Does this mean that the `6` in `pyramid_down<6>` is wrong?

**Comment by davisking** [comment]
It's fine.  You are just upsampling the images in python before processing but not in C++.  Upsample the images in C++ before giving them to the detector.

**Comment by florian** [comment]
Thanks a lot for the quick reply!

Do you mean just calling `pyramid_up(cimg)`? How often does this method need to be called to have the same upsampling as in Python?
Is there any good way to call this method when working with OpenCV images?

**Comment by davisking** [comment]
Yes, that's what I mean.  How much depends on what the python code did.  That message you posted makes it sound like you should call it 2 times.  

If you want to use OpenCV things you should look at the OpenCV documentation.

**Comment by florian** [comment]
Great, thanks! That seems to have improved things.

The bounding box is now off, I guess this is because we upsampled the image.
Sorry if this is another very simple question, but by how much does `pyramid_up` increase the image size? It seems to be image dependent, calling the method twice sometimes increases the width by 4.04, sometimes by 4.055. I searched for a while but didn't find anything on this.
